File: web/src/lib/server/members.ts
// Admin member-management BFF client (spec 008 T10). The SvelteKit admin tier is a **BFF, not a
// re-implementation** (P4 / ADR-0026): it owns session + presentation and CALLS the Rust Worker
// (`/api/admin/members/*`, T09) with the server-to-server shared secret + the WebAuthn-verified acting
// admin id — it never re-does the crypto/validation/normalization the core owns.
//
// FUNCTIONAL-CORE / IMPERATIVE-SHELL (mirrors `kv-challenge-store.ts`): this module is PURE — types
// derived by hand from the frozen `api/openapi.yaml` (plan §6: hand-rolled-but-derived for v1; no TS
// codegen yet), the `MembersClient` port, the real `WorkerMembersClient` fetch adapter, a seedable
// `InMemoryMembersClient` dev/e2e fake, and the fail-closed `selectMembersClient` decision. It imports
// NO SvelteKit-virtual modules, so it is unit-testable under the bare Vitest config; the imperative
// shell (`members-deps.ts`) supplies the Worker base URL + secret (from `$env/dynamic/private`), the
// in-memory fallback singleton, and `dev`.
//
// This file lives under `src/lib/server/` (SvelteKit never bundles it to the client), so the shared
// secret can never leak into client code. The wire TYPES live in the client-safe `$lib/members-types`
// (so components can type `data.members`); this module imports + re-exports them.

import type {
	AuditEntry,
	AuditField,
	DetailOutcome,
	EditMemberRequest,
	EditOutcome,
	IssueMemberRequest,
	IssueOutcome,
	ListParams,
	MemberDetail,
	MembersClient,
	MemberSummary,
	OnboardingStatus,
	RegenerateOutcome,
	Role,
	SeedMemberInput,
} from '../members-types';

export type {
	AuditEntry,
	AuditField,
	DetailOutcome,
	EditMemberRequest,
	EditOutcome,
	IssuableRole,
	IssueMemberRequest,
	IssueOutcome,
	ListParams,
	MemberDetail,
	MembersClient,
	MemberSummary,
	OnboardingStatus,
	RegenerateOutcome,
	Role,
	SeedMemberInput,
} from '../members-types';

// ── The real Worker adapter (ADR-0026): server-to-server bearer secret + the asserted X-Admin-Id ──

interface ErrorBody {
	readonly error_code: string;
}

export class WorkerMembersClient implements MembersClient {
	constructor(
		private readonly baseUrl: string,
		private readonly secret: string,
	) {}

	private url(path: string): string {
		return `${this.baseUrl.replace(/\/$/, '')}${path}`;
	}

	private headers(adminId: string, json = false): Record<string, string> {
		const h: Record<string, string> = {
			authorization: `Bearer ${this.secret}`,
			'x-admin-id': adminId,
		};
		if (json) h['content-type'] = 'application/json';
		return h;
	}

	/** A non-outcome status (401 BFF-misconfig, or anything unexpected) is an operator fault. Throw a
	 *  value-free error — never echo the response body onto the BFF's own error/log path (P2). */
	private fail(res: Response): Error {
		return new Error(`admin member Worker call failed: HTTP ${res.status}`);
	}

	async list(adminId: string, params: ListParams): Promise<MemberSummary[]> {
		const qs = new URLSearchParams();
		if (params.search) qs.set('search', params.search);
		if (params.role) qs.set('role', params.role);
		if (params.status) qs.set('status', params.status);
		const q = qs.toString();
		const res = await fetch(this.url(`/api/admin/members${q ? `?${q}` : ''}`), {
			headers: this.headers(adminId),
		});
		if (res.status === 200) return ((await res.json()) as { members: MemberSummary[] }).members;
		throw this.fail(res);
	}

	async issue(adminId: string, req: IssueMemberRequest): Promise<IssueOutcome> {
		const res = await fetch(this.url('/api/admin/members'), {
			method: 'POST',
			headers: this.headers(adminId, true),
			body: JSON.stringify(req),
		});
		if (res.status === 201) {
			const b = (await res.json()) as { member: MemberSummary; onboarding_code?: string; code_expires_at: number };
			return { kind: 'issued', member: b.member, onboarding_code: b.onboarding_code, code_expires_at: b.code_expires_at };
		}
		if (res.status === 409) {
			const b = (await res.json()) as { existing: MemberSummary };
			return { kind: 'duplicate', existing: b.existing };
		}
		if (res.status === 400) return { kind: 'rejected', code: ((await res.json()) as ErrorBody).error_code };
		if (res.status === 503) return { kind: 'group_key_missing' };
		throw this.fail(res);
	}

	async detail(adminId: string, id: string): Promise<DetailOutcome> {
		const res = await fetch(this.url(`/api/admin/members/${encodeURIComponent(id)}`), {
			headers: this.headers(adminId),
		});
		if (res.status === 200) return { kind: 'detail', detail: (await res.json()) as MemberDetail };
		if (res.status === 404) return { kind: 'not_found' };
		if (res.status === 503) return { kind: 'group_key_missing' };
		throw this.fail(res);
	}

	async edit(adminId: string, id: string, req: EditMemberRequest): Promise<EditOutcome> {
		const res = await fetch(this.url(`/api/admin/members/${encodeURIComponent(id)}`), {
			method: 'PATCH',
			headers: this.headers(adminId, true),
			body: JSON.stringify(req),
		});
		if (res.status === 200) return { kind: 'updated', detail: (await res.json()) as MemberDetail };
		if (res.status === 409) return { kind: 'stale' };
		if (res.status === 400) return { kind: 'rejected', code: ((await res.json()) as ErrorBody).error_code };
		if (res.status === 404) return { kind: 'not_found' };
		if (res.status === 503) return { kind: 'group_key_missing' };
		throw this.fail(res);
	}

	async regenerateCode(adminId: string, id: string): Promise<RegenerateOutcome> {
		const res = await fetch(this.url(`/api/admin/members/${encodeURIComponent(id)}/regenerate-code`), {
			method: 'POST',
			headers: this.headers(adminId),
		});
		if (res.status === 200) {
			const b = (await res.json()) as { onboarding_code?: string; code_expires_at: number };
			return { kind: 'regenerated', onboarding_code: b.onboarding_code, code_expires_at: b.code_expires_at };
		}
		if (res.status === 404) return { kind: 'not_found' };
		throw this.fail(res);
	}

	async auditLog(adminId: string, params: { memberId?: string }): Promise<AuditEntry[]> {
		const q = params.memberId ? `?member_id=${encodeURIComponent(params.memberId)}` : '';
		const res = await fetch(this.url(`/api/admin/audit-log${q}`), { headers: this.headers(adminId) });
		if (res.status === 200) return ((await res.json()) as { entries: AuditEntry[] }).entries;
		throw this.fail(res);
	}
}

// ── In-memory fake (dev + e2e). Mirrors the Worker's contract — the show-once code, the audited detail
//    read, the duplicate-phone 409 link, optimistic-concurrency staleness — so the UI is fully driveable
//    locally with no Worker/account. It holds plaintext (it is the dev backend, never a deploy target —
//    the fail-closed `selectMembersClient` refuses it outside dev). ──

interface StoredMember {
	member_id: string;
	name: string;
	phone: string;
	address: string;
	roles: Role[];
	onboarding_status: OnboardingStatus;
	updated_at: number;
	onboarding_code: string;
	code_expires_at: number;
}

const CODE_TTL_SECS = 72 * 60 * 60;

export class InMemoryMembersClient implements MembersClient {
	private readonly members: StoredMember[] = [];
	private readonly audit: AuditEntry[] = [];
	/** Monotonic non-colliding `updated_at`/timestamp source — so a stale token always differs and the
	 *  optimistic-concurrency reject is exercisable even within one wall-clock second. */
	private stamp = Math.floor(Date.now() / 1000);

	private nextStamp(): number {
		this.stamp += 1;
		return this.stamp;
	}

	private mintCode(): string {
		return `BNDL-${globalThis.crypto.randomUUID().replace(/-/g, '').slice(0, 8).toUpperCase()}`;
	}

	private summary(m: StoredMember): MemberSummary {
		return { member_id: m.member_id, name: m.name, roles: m.roles, onboarding_status: m.onboarding_status };
	}

	private record(adminId: string, memberId: string, fields: AuditField[]): void {
		this.audit.push({
			timestamp: this.nextStamp(),
			admin_id: adminId,
			member_id: memberId,
			fields,
			request_id: globalThis.crypto.randomUUID(),
		});
	}

	/** Light validation mirroring the core's outcomes (the real normalization/hash lives in the core). */
	private validate(name: string, phone: string, address: string, roles: string[]): { code: string } | null {
		if (roles.length === 0) return { code: 'ADMIN_MEMBER_ROLES_REQUIRED' };
		if (roles.some((r) => r === 'admin')) return { code: 'ADMIN_MEMBER_ROLE_FORBIDDEN' };
		if (address.trim() === '') return { code: 'ADMIN_MEMBER_ADDRESS_INVALID' };
		const digits = phone.replace(/[^0-9]/g, '');
		if (!/^\+?[0-9 ()-]+$/.test(phone) || digits.length < 7) return { code: 'ADMIN_MEMBER_PHONE_INVALID' };
		return null;
	}

	/** Normalize a phone the way the duplicate check compares (digits only — the fake's stand-in for
	 *  the core's E.164 `normalize_phone`). */
	private normalize(phone: string): string {
		return phone.replace(/[^0-9]/g, '');
	}

	seed(input: SeedMemberInput): MemberSummary {
		const m: StoredMember = {
			member_id: input.member_id ?? globalThis.crypto.randomUUID(),
			name: input.name,
			phone: input.phone,
			address: input.address,
			roles: input.roles ?? ['rider'],
			onboarding_status: input.onboarding_status ?? 'issued_not_onboarded',
			updated_at: this.nextStamp(),
			onboarding_code: this.mintCode(),
			code_expires_at: Math.floor(Date.now() / 1000) + CODE_TTL_SECS,
		};
		this.members.push(m);
		return this.summary(m);
	}

	async list(_adminId: string, params: ListParams): Promise<MemberSummary[]> {
		const search = params.search?.trim().toLowerCase();
		return this.members
			.filter((m) => (search ? m.name.toLowerCase().includes(search) : true))
			.filter((m) => (params.role ? m.roles.includes(params.role) : true))
			.filter((m) => (params.status ? m.onboarding_status === params.status : true))
			.map((m) => this.summary(m));
	}

	async issue(adminId: string, req: IssueMemberRequest): Promise<IssueOutcome> {
		const rejected = this.validate(req.name, req.phone, req.address, req.roles);
		if (rejected) return { kind: 'rejected', code: rejected.code };

		const existing = this.members.find((m) => this.normalize(m.phone) === this.normalize(req.phone));
		if (existing) {
			// Duplicate disclosure is an admin-only, I5-audited read (name only — R9).
			this.record(adminId, existing.member_id, ['name']);
			return { kind: 'duplicate', existing: this.summary(existing) };
		}

		const m: StoredMember = {
			member_id: globalThis.crypto.randomUUID(),
			name: req.name,
			phone: req.phone,
			address: req.address,
			roles: [...req.roles],
			onboarding_status: 'issued_not_onboarded',
			updated_at: this.nextStamp(),
			onboarding_code: this.mintCode(),
			code_expires_at: Math.floor(Date.now() / 1000) + CODE_TTL_SECS,
		};
		this.members.push(m);
		return { kind: 'issued', member: this.summary(m), onboarding_code: m.onboarding_code, code_expires_at: m.code_expires_at };
	}

	async detail(adminId: string, id: string): Promise<DetailOutcome> {
		const m = this.members.find((x) => x.member_id === id);
		if (!m) return { kind: 'not_found' };
		this.record(adminId, m.member_id, ['name', 'phone', 'address']);
		return {
			kind: 'detail',
			detail: {
				member_id: m.member_id,
				name: m.name,
				phone: m.phone,
				address: m.address,
				roles: m.roles,
				onboarding_status: m.onboarding_status,
				updated_at: m.updated_at,
			},
		};
	}

	async edit(adminId: string, id: string, req: EditMemberRequest): Promise<EditOutcome> {
		const m = this.members.find((x) => x.member_id === id);
		if (!m) return { kind: 'not_found' };
		if (m.updated_at !== req.expected_updated_at) return { kind: 'stale' };

		const name = req.name ?? m.name;
		const phone = req.phone ?? m.phone;
		const address = req.address ?? m.address;
		const roles = req.roles ?? m.roles;
		const rejected = this.validate(name, phone, address, roles);
		if (rejected) return { kind: 'rejected', code: rejected.code };
		// A phone change onto a number already enrolled is an opaque store conflict on the real Worker
		// (T09 register edit-into-duplicate carry-forward); the fake mirrors the happy path only.

		m.name = name;
		m.phone = phone;
		m.address = address;
		m.roles = [...roles];
		m.updated_at = this.nextStamp();
		// PATCH returns MemberDetail (PII) → an audited read (x-requires-audit, contract).
		this.record(adminId, m.member_id, ['name', 'phone', 'address']);
		return {
			kind: 'updated',
			detail: {
				member_id: m.member_id,
				name: m.name,
				phone: m.phone,
				address: m.address,
				roles: m.roles,
				onboarding_status: m.onboarding_status,
				updated_at: m.updated_at,
			},
		};
	}

	async regenerateCode(_adminId: string, id: string): Promise<RegenerateOutcome> {
		const m = this.members.find((x) => x.member_id === id);
		if (!m) return { kind: 'not_found' };
		m.onboarding_code = this.mintCode();
		m.code_expires_at = Math.floor(Date.now() / 1000) + CODE_TTL_SECS;
		return { kind: 'regenerated', onboarding_code: m.onboarding_code, code_expires_at: m.code_expires_at };
	}

	async auditLog(_adminId: string, params: { memberId?: string }): Promise<AuditEntry[]> {
		return params.memberId ? this.audit.filter((e) => e.member_id === params.memberId) : [...this.audit];
	}
}

// ── Fail-closed selection (mirrors `selectChallengeStore`) ──

/**
 * Pick the members client for a request. Returns the real `WorkerMembersClient` (ADR-0026) when BOTH the
 * Worker base URL and the shared secret are configured (the deployed edge). When either is absent it
 * falls back to the in-memory fake ONLY when `allowInMemoryFallback` is true (dev/e2e) — otherwise it
 * **fails closed** by throwing, so a production build with no Worker binding refuses to serve a fake
 * member roster rather than degrade silently.
 *
 * Pure (no SvelteKit-virtual imports) so it is unit-testable; the imperative shell (`members-deps.ts`)
 * supplies the env values, the fallback singleton, and `dev`.
 */
export function selectMembersClient(
	baseUrl: string | undefined,
	secret: string | undefined,
	fallback: MembersClient,
	allowInMemoryFallback: boolean,
): MembersClient {
	if (baseUrl && secret) return new WorkerMembersClient(baseUrl, secret);
	if (allowInMemoryFallback) return fallback;
	throw new Error(
		'Admin member backend unavailable: ADMIN_WORKER_BASE / ADMIN_API_SECRET are not configured. ' +
			'Refusing the in-memory fallback outside dev — it serves an empty/fake member roster and would ' +
			'mask a deploy misconfiguration (ADR-0026).',
	);
}

File: server/src/runtime/members.rs
//! The deployable admin member-management routes (spec 008 **T09**) — wasm-only.
//!
//! `/api/admin/members/*` + `/api/admin/audit-log`: the producer side of the closed-group model
//! (issue / list / detail / edit / regenerate-code / audit-log). Composes the **real** core
//! [`MemberService`] over the **real** [`PgMemberStore`] (P4 — issuance logic single-sourced), with a
//! live getrandom-backed CSPRNG injected into [`RngSecretSource`] (ADR-0021: the Worker — not the
//! core — holds randomness) and the per-Group key loaded + unwrapped **per request** from the KEK
//! (no long-lived plaintext key cached in the DO — see `DEFERRED.md` → T09).
//!
//! ## Trust model (ADR-0026)
//!
//! ADR-0017 put admin WebAuthn verification in the SvelteKit tier; the Worker has no admin-session
//! notion. So the WebAuthn-verified SvelteKit **BFF** calls these routes with a server-to-server
//! shared secret (`ADMIN_API_SECRET`) and asserts the verified acting admin id (`X-Admin-Id`). The
//! [`admin_guard`] checks the secret in **constant time** and **fails closed** without it (401); a
//! missing binding is a 500 (misconfig), a missing/bad admin id a 400. `group_id` is the single-
//! install `GROUP_ID` binding (the RLS tenant). The real SvelteKit→Worker call is wired at T10.
//!
//! ## Privacy spine in the shell (P2/R10/I5)
//!
//! - Inbound `name`/`address`/`phone` are wrapped into the **tainted** [`MemberName`]/[`Address`]/
//!   [`PhoneNumber`] (no `Debug`/`Serialize`) the moment the body is parsed; they are never logged and
//!   never echoed — every rejection returns a stable, value-free [`ErrorBody`](err_code) code (R10).
//! - Every admin response body is serialized through the sealed [`admin_response_body`] seam (the I5
//!   gate): the decrypted-PII detail can only reach the wire as a `PiiDisclosure<MemberDetailView>`
//!   minted after the audit committed, and the other bodies are the blessed PII-free envelope views.
//!   The Worker hand-rolls **no** member-PII JSON.

use boundless_auth::UnixSeconds;
use boundless_domain::{Address, MemberId, MemberName, PhoneNumber, Role};
use boundless_server_core::{
    admin_response_body, issuable_roles, AuditLogView, AuditedResponse, DetailRead,
    DuplicatePhoneLinkView, EditMemberInput, EditMemberOutcome, IssueMemberInput,
    IssueMemberOutcome, MemberConfig, MemberError, MemberIssuedView, MemberListView, MemberService,
    RegenerateCodeView, RegenerateOutcome, RngSecretSource,
};
use boundless_server_store::{ensure_least_privilege, PgMemberStore};
use rand_core::{CryptoRng, RngCore};
use serde::Deserialize;
use uuid::Uuid;
use worker::{Env, Headers, Request, Response, Result, RouteContext};

use super::pg::{connect_pg, load_group_id, load_hmac_key, load_kek, WorkerClock};

// ===== the live Worker CSPRNG (ADR-0021) ==========================================================

/// A getrandom-backed CSPRNG, injected into [`RngSecretSource`] so the core stays randomness-free
/// (ADR-0021). getrandom's `wasm_js` backend (Web Crypto) provides workerd entropy. The `.expect()` on
/// each draw is **load-bearing**: ignoring the error could yield all-zero bytes — a catastrophic
/// nonce/key reuse (R1) — so a getrandom failure must panic (fail-closed: the request 500s) rather than
/// ever produce predictable output.
struct GetrandomRng;

impl RngCore for GetrandomRng {
    fn next_u32(&mut self) -> u32 {
        let mut b = [0u8; 4];
        getrandom::fill(&mut b).expect("workerd getrandom (wasm_js) must not fail");
        u32::from_ne_bytes(b)
    }
    fn next_u64(&mut self) -> u64 {
        let mut b = [0u8; 8];
        getrandom::fill(&mut b).expect("workerd getrandom (wasm_js) must not fail");
        u64::from_ne_bytes(b)
    }
    fn fill_bytes(&mut self, dst: &mut [u8]) {
        getrandom::fill(dst).expect("workerd getrandom (wasm_js) must not fail");
    }
}

// The `CryptoRng` marker asserts the generator is cryptographically secure (getrandom is).
impl CryptoRng for GetrandomRng {}

/// The composed engine the admin routes drive: the real `PgMemberStore` over Hyperdrive + the live
/// CSPRNG + server time + the per-instance config (HMAC + KEK).
type AdminService = MemberService<PgMemberStore, RngSecretSource<GetrandomRng>, WorkerClock>;

/// Connect over Hyperdrive, run the W2 least-privilege guard (fail closed), and assemble the member
/// engine over the real `PgMemberStore` with the live CSPRNG, the HMAC key, and the KEK. Per-request —
/// a fresh connection + a fresh per-request Group-key unwrap (inside the service), the Hyperdrive pooler
/// pattern. Any failure (binding missing, too-privileged role, bad config) maps to a value-free error.
async fn build_member_service(env: &Env) -> Result<AdminService> {
    let client = connect_pg(env).await?;
    // W2 (sec-audit): refuse a superuser / BYPASSRLS role — RLS would silently not apply → cross-tenant
    // PII. The detail is value-free; the caller sees a generic 500.
    ensure_least_privilege(&client)
        .await
        .map_err(|_| worker::Error::RustError("db role misconfigured".into()))?;
    let group_id = load_group_id(env)?;
    let config = MemberConfig {
        hmac_key: load_hmac_key(env)?,
        kek: load_kek(env)?,
    };
    Ok(MemberService::new(
        PgMemberStore::new(client, group_id),
        RngSecretSource::new(GetrandomRng),
        WorkerClock,
        config,
    ))
}

// ===== the ADR-0026 shared-secret admin gate =====================================================

/// The shared-secret + asserted-admin-id gate (ADR-0026). Returns the verified acting [`MemberId`], or
/// an early reject `Response` (the inner `Err`). Fails **closed**: a missing `ADMIN_API_SECRET` binding
/// → 500 (misconfig); a missing/wrong bearer → 401; a missing/bad `X-Admin-Id` → 400. The secret
/// compare is constant-time (no content-timing leak). The outer `Result` is for genuinely-fallible
/// header/Response construction.
fn admin_guard(req: &Request, env: &Env) -> Result<core::result::Result<MemberId, Response>> {
    let Ok(expected) = env.var("ADMIN_API_SECRET").map(|v| v.to_string()) else {
        // Misconfigured deploy — fail closed with a value-free 500 (never reveal the binding state).
        return Ok(Err(Response::error("admin api not configured", 500)?));
    };
    let presented = req
        .headers()
        .get("authorization")?
        .and_then(|h| h.strip_prefix("Bearer ").map(str::to_string))
        .unwrap_or_default();
    if !constant_time_eq(presented.as_bytes(), expected.as_bytes()) {
        return Ok(Err(err_code("ADMIN_UNAUTHORIZED", 401)?));
    }
    // The BFF-asserted acting admin id (trusted because the shared secret verified) — the I5 audit actor.
    match req
        .headers()
        .get("x-admin-id")?
        .and_then(|s| Uuid::parse_str(s.trim()).ok())
    {
        Some(u) => Ok(Ok(MemberId::from_uuid(u))),
        None => Ok(Err(err_code("ADMIN_BAD_REQUEST", 400)?)),
    }
}

/// Constant-time byte-equality (no early exit on content). The length check leaks only the length of a
/// high-entropy shared secret, not its content — the timing-sensitive part.
fn constant_time_eq(a: &[u8], b: &[u8]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    let mut diff = 0u8;
    for (x, y) in a.iter().zip(b.iter()) {
        diff |= x ^ y;
    }
    diff == 0
}

/// Run [`admin_guard`] and bind the acting admin id, or short-circuit the handler with the reject.
macro_rules! guard_admin {
    ($req:expr, $env:expr) => {
        match admin_guard(&$req, &$env)? {
            Ok(id) => id,
            Err(resp) => return Ok(resp),
        }
    };
}

// ===== response helpers ===========================================================================

/// Emit a body already serialized through the sealed [`admin_response_body`] seam (the I5 gate) with a
/// JSON content-type + status. The `R: AuditedResponse + Serialize` bound means a bare PII DTO cannot
/// be sent here — only a blessed envelope or a `PiiDisclosure` (which an audit minted).
fn audited_body<R: AuditedResponse + serde::Serialize>(view: &R, status: u16) -> Result<Response> {
    let body = admin_response_body(view)
        .map_err(|_| worker::Error::RustError("serialize admin body".into()))?;
    let headers = Headers::new();
    headers.set("content-type", "application/json")?;
    Ok(Response::ok(body)?
        .with_status(status)
        .with_headers(headers))
}

/// A PII-free [`ErrorBody`](https://) envelope — only a stable code, never the submitted value (P2/R10).
fn err_code(code: &str, status: u16) -> Result<Response> {
    Ok(Response::from_json(&serde_json::json!({ "error_code": code }))?.with_status(status))
}

/// The HTTP status for a validation rejection. Bad input → 400; no Group key → 503 (operator).
fn status_for(e: MemberError) -> u16 {
    match e {
        MemberError::PhoneInvalid | MemberError::AddressInvalid | MemberError::RolesRequired => 400,
        MemberError::GroupKeyMissing => 503,
    }
}

/// Mint a server-minted opaque request-correlation id (never client-echoed — it lands in the audit
/// row, so a client-supplied value would be an injection point). 16 random bytes, lowercase hex.
fn mint_request_id() -> String {
    let mut b = [0u8; 16];
    getrandom::fill(&mut b).expect("workerd getrandom (wasm_js) must not fail");
    let mut s = String::with_capacity(32);
    for byte in b {
        s.push(char::from_digit((byte >> 4) as u32, 16).unwrap_or('0'));
        s.push(char::from_digit((byte & 0x0f) as u32, 16).unwrap_or('0'));
    }
    s
}

/// The path `:id` segment → a [`MemberId`], or `None` on a malformed uuid.
fn path_member_id(ctx: &RouteContext<()>) -> Option<MemberId> {
    Uuid::parse_str(ctx.param("id")?.trim())
        .ok()
        .map(MemberId::from_uuid)
}

/// The optional `?member_id=` audit-log filter.
fn query_member_id(req: &Request) -> Option<MemberId> {
    let url = req.url().ok()?;
    let v = url
        .query_pairs()
        .find(|(k, _)| k == "member_id")
        .map(|(_, v)| v.into_owned())?;
    Uuid::parse_str(v.trim()).ok().map(MemberId::from_uuid)
}

// ===== request bodies =============================================================================

/// `POST /api/admin/members` body (the contract `IssueMemberRequest`). PII is wrapped into tainted
/// types immediately on use; `roles` deserialize to the domain `Role` so an `admin` is rejected
/// server-side by the core `issuable_roles` (I11), not merely by the wire enum.
#[derive(Deserialize)]
struct IssueWire {
    name: String,
    phone: String,
    address: String,
    roles: Vec<Role>,
}

/// `PATCH /api/admin/members/{id}` body (the contract `EditMemberRequest`). Every PII field optional.
#[derive(Deserialize)]
struct EditWire {
    name: Option<String>,
    phone: Option<String>,
    address: Option<String>,
    roles: Option<Vec<Role>>,
    expected_updated_at: i64,
}

// ===== the routes =================================================================================

/// `GET /api/admin/members` — the PII-free member list (AC8; not an audited read). The `search`/`role`/
/// `status` query filters are accepted but **not yet applied** (the core `list_members` takes no filter
/// — server-side search/filter is deferred, see `DEFERRED.md` → T07/T10).
pub(crate) async fn list(req: Request, ctx: RouteContext<()>) -> Result<Response> {
    let _admin = guard_admin!(req, ctx.env);
    let mut svc = match build_member_service(&ctx.env).await {
        Ok(s) => s,
        Err(_) => return Response::error("internal", 500),
    };
    match svc.list_members().await {
        Ok(Ok(members)) => audited_body(&MemberListView::new(members), 200),
        Ok(Err(e)) => err_code(e.error_code(), status_for(e)),
        Err(_) => Response::error("internal", 500),
    }
}

/// `POST /api/admin/members` — issue a Rider/Driver member + mint a one-time Onboarding Code
/// (AC1/AC5). A phone collision surfaces-and-links the existing member (I5-audited in the store).
pub(crate) async fn issue(mut req: Request, ctx: RouteContext<()>) -> Result<Response> {
    let admin = guard_admin!(req, ctx.env);
    let Ok(body) = req.json::<IssueWire>().await else {
        return err_code("ADMIN_BAD_REQUEST", 400);
    };
    // Reject the Admin role server-side (I11/AC10) — the single-source core conversion.
    let roles = match issuable_roles(&body.roles) {
        Ok(r) => r,
        Err(_) => return err_code("ADMIN_MEMBER_ROLE_FORBIDDEN", 400),
    };
    let input = IssueMemberInput {
        name: MemberName::new(body.name),
        phone: PhoneNumber::new(body.phone),
        address: Address::new(body.address),
        roles,
    };
    let mut svc = match build_member_service(&ctx.env).await {
        Ok(s) => s,
        Err(_) => return Response::error("internal", 500),
    };
    match svc.issue_member(admin, input, mint_request_id()).await {
        Ok(IssueMemberOutcome::Issued {
            member,
            onboarding_code,
            code_expires_at,
        }) => audited_body(
            &MemberIssuedView::new(member, &onboarding_code, code_expires_at),
            201,
        ),
        Ok(IssueMemberOutcome::DuplicatePhone { existing }) => {
            audited_body(&DuplicatePhoneLinkView::new(existing), 409)
        }
        Ok(IssueMemberOutcome::Rejected(e)) => err_code(e.error_code(), status_for(e)),
        Err(_) => Response::error("internal", 500),
    }
}

/// `GET /api/admin/members/{id}` — the audited PII detail read (AC7). The audit row is written
/// atomically with the ciphertext SELECT in the store (I5/§7); the decrypted view reaches the wire only
/// as the `PiiDisclosure` the read minted.
pub(crate) async fn detail(req: Request, ctx: RouteContext<()>) -> Result<Response> {
    let admin = guard_admin!(req, ctx.env);
    let Some(id) = path_member_id(&ctx) else {
        return err_code("ADMIN_MEMBER_NOT_FOUND", 404);
    };
    let mut svc = match build_member_service(&ctx.env).await {
        Ok(s) => s,
        Err(_) => return Response::error("internal", 500),
    };
    match svc.read_detail(admin, id, mint_request_id()).await {
        Ok(DetailRead::Detail(disclosure)) => audited_body(&*disclosure, 200),
        Ok(DetailRead::NotFound) => err_code("ADMIN_MEMBER_NOT_FOUND", 404),
        Ok(DetailRead::GroupKeyMissing) => err_code("ADMIN_GROUP_KEY_MISSING", 503),
        Err(_) => Response::error("internal", 500),
    }
}

/// `PATCH /api/admin/members/{id}` — edit under optimistic concurrency, then return the updated detail
/// (an audited read — AC11). The edit itself is not an audited read; the read-back writes the audit row.
pub(crate) async fn edit(mut req: Request, ctx: RouteContext<()>) -> Result<Response> {
    let admin = guard_admin!(req, ctx.env);
    let Some(id) = path_member_id(&ctx) else {
        return err_code("ADMIN_MEMBER_NOT_FOUND", 404);
    };
    let Ok(body) = req.json::<EditWire>().await else {
        return err_code("ADMIN_BAD_REQUEST", 400);
    };
    let roles = match body.roles {
        Some(rs) => match issuable_roles(&rs) {
            Ok(r) => Some(r),
            Err(_) => return err_code("ADMIN_MEMBER_ROLE_FORBIDDEN", 400),
        },
        None => None,
    };
    let input = EditMemberInput {
        name: body.name.map(MemberName::new),
        phone: body.phone.map(PhoneNumber::new),
        address: body.address.map(Address::new),
        roles,
        expected_updated_at: UnixSeconds::new(body.expected_updated_at),
    };
    let mut svc = match build_member_service(&ctx.env).await {
        Ok(s) => s,
        Err(_) => return Response::error("internal", 500),
    };
    match svc.edit_member(id, input, mint_request_id()).await {
        Ok(EditMemberOutcome::Updated) => {
            // Return the updated detail — an audited read (the audit row is for this read-back).
            match svc.read_detail(admin, id, mint_request_id()).await {
                Ok(DetailRead::Detail(disclosure)) => audited_body(&*disclosure, 200),
                Ok(DetailRead::NotFound) => err_code("ADMIN_MEMBER_NOT_FOUND", 404),
                Ok(DetailRead::GroupKeyMissing) => err_code("ADMIN_GROUP_KEY_MISSING", 503),
                Err(_) => Response::error("internal", 500),
            }
        }
        Ok(EditMemberOutcome::Stale) => err_code("ADMIN_MEMBER_EDIT_STALE", 409),
        Ok(EditMemberOutcome::Rejected(e)) => err_code(e.error_code(), status_for(e)),
        Err(_) => Response::error("internal", 500),
    }
}

/// `POST /api/admin/members/{id}/regenerate-code` — mint a fresh Onboarding Code, supersede the prior
/// live one atomically (AC6). Not an audited read (no name/phone/address is disclosed).
pub(crate) async fn regenerate(req: Request, ctx: RouteContext<()>) -> Result<Response> {
    let _admin = guard_admin!(req, ctx.env);
    let Some(id) = path_member_id(&ctx) else {
        return err_code("ADMIN_MEMBER_NOT_FOUND", 404);
    };
    let mut svc = match build_member_service(&ctx.env).await {
        Ok(s) => s,
        Err(_) => return Response::error("internal", 500),
    };
    match svc.regenerate_onboarding_code(id).await {
        Ok(RegenerateOutcome::Regenerated {
            onboarding_code,
            code_expires_at,
        }) => audited_body(
            &RegenerateCodeView::new(&onboarding_code, code_expires_at),
            200,
        ),
        Ok(RegenerateOutcome::NotFound) => err_code("ADMIN_MEMBER_NOT_FOUND", 404),
        Err(_) => Response::error("internal", 500),
    }
}

/// `GET /api/admin/audit-log` — the I5 PII-read audit log (AC9), field names only, optionally filtered
/// to one member. Not itself an audited read (no values).
pub(crate) async fn audit_log(req: Request, ctx: RouteContext<()>) -> Result<Response> {
    let _admin = guard_admin!(req, ctx.env);
    let member = query_member_id(&req);
    let mut svc = match build_member_service(&ctx.env).await {
        Ok(s) => s,
        Err(_) => return Response::error("internal", 500),
    };
    match svc.read_audit_log(member).await {
        Ok(entries) => audited_body(&AuditLogView::new(entries), 200),
        Err(_) => Response::error("internal", 500),
    }
}
